diff -ruN STM32Cube_FW_F4_V1.13.0/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c STM32Cube_FW_F4_V1.13.0-ap/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c
--- STM32Cube_FW_F4_V1.13.0/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c	2016-06-27 17:02:18.000000000 -0400
+++ STM32Cube_FW_F4_V1.13.0-ap/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c	2016-07-30 16:05:52.778517078 -0400
@@ -57,6 +57,7 @@
 
 /* Includes ------------------------------------------------------------------*/
 #include "stm32f4xx_hal.h"
+#include <string.h>
 
 /** @addtogroup STM32F4xx_LL_USB_DRIVER
   * @{
@@ -783,7 +784,9 @@
     count32b =  (len + 3U) / 4U;
     for (i = 0U; i < count32b; i++, src += 4U)
     {
-      USBx_DFIFO(ch_ep_num) = *((__packed uint32_t *)src);
+      uint32_t val;
+      memcpy(&val, src, sizeof(val));
+      USBx_DFIFO(ch_ep_num) = val;
     }
   }
   return HAL_OK;
diff -ruN STM32Cube_FW_F4_V1.13.0/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Inc/usbd_cdc.h STM32Cube_FW_F4_V1.13.0-ap/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Inc/usbd_cdc.h
--- STM32Cube_FW_F4_V1.13.0/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Inc/usbd_cdc.h	2016-06-27 14:19:56.000000000 -0400
+++ STM32Cube_FW_F4_V1.13.0-ap/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Inc/usbd_cdc.h	2016-07-30 16:05:47.746581712 -0400
@@ -104,7 +104,7 @@
   int8_t (* DeInit)        (void);
   int8_t (* Control)       (uint8_t, uint8_t * , uint16_t);   
   int8_t (* Receive)       (uint8_t *, uint32_t *);  
-
+  void   (* TxCompleted)   (void);
 }USBD_CDC_ItfTypeDef;
 
 
@@ -159,6 +159,8 @@
 uint8_t  USBD_CDC_ReceivePacket      (USBD_HandleTypeDef *pdev);
 
 uint8_t  USBD_CDC_TransmitPacket     (USBD_HandleTypeDef *pdev);
+
+uint8_t  USBD_CDC_CheckTxBusy        (USBD_HandleTypeDef *pdev);
 /**
   * @}
   */ 
diff -ruN STM32Cube_FW_F4_V1.13.0/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Src/usbd_cdc.c STM32Cube_FW_F4_V1.13.0-ap/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Src/usbd_cdc.c
--- STM32Cube_FW_F4_V1.13.0/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Src/usbd_cdc.c	2016-06-27 14:19:56.000000000 -0400
+++ STM32Cube_FW_F4_V1.13.0-ap/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Src/usbd_cdc.c	2016-07-30 16:19:42.596190111 -0400
@@ -669,6 +669,8 @@
   {
     
     hcdc->TxState = 0;
+    
+    ((USBD_CDC_ItfTypeDef *)pdev->pUserData)->TxCompleted();
 
     return USBD_OK;
   }
@@ -873,6 +875,19 @@
   }
 }
 
+uint8_t  USBD_CDC_CheckTxBusy(USBD_HandleTypeDef *pdev)
+{      
+  USBD_CDC_HandleTypeDef   *hcdc = (USBD_CDC_HandleTypeDef*) pdev->pClassData;
+  
+  if(pdev->pClassData != NULL)
+  {
+    return hcdc->TxState == 0 ? USBD_OK : USBD_BUSY;
+  }
+  else
+  {
+    return USBD_FAIL;
+  }
+}
 
 /**
   * @brief  USBD_CDC_ReceivePacket
